File: Linked Lists/LinkedList.py
from Node import Node

class LinkedList:

    def __init__(self) -> None:
        self.head = Node()
        self.tail = Node()
        self.size = 0

    def append(self, val):

        # OBJECTIVE: Add a new node to the end of the list

        # Create a new node
        newNode = Node()
        newNode.val = val

        # If head is by itself, add newNode after it
        if self.head.next == None:

            # Connect newNode to head and tail nodes
            newNode.prev = self.head
            newNode.next = self.tail

            # Connect head and tail nodes to newNode
            self.head.next = newNode
            self.tail.prev = newNode
        
        # If not, add newNode before tail node
        else:

            # Get node before tail
            lastNode = self.tail.prev

            # Connect lastNode to newNode and vice versa
            lastNode.next = newNode
            newNode.prev = lastNode

            # Connect newNode to tail node and vice versa
            newNode.next = self.tail
            self.tail.prev = newNode

        self.size += 1

    def getHead(self):

        # OBJECTIVE: Return head node if it exists

        # If there's no head node, exit function
        if self.head.next == None:
            return

        return self.head.next

    def getSize(self):

        # OBJECTIVE: Get linked list size
        return self.size

    def printList(self):

        # OBJECTIVE: Print linked list from head

        # If linked list is empty, exit function
        if self.head.next == None:
            print("Linked list is empty")
            return

        # Get head node
        curNode = self.head.next

        # Traverse linked list
        while curNode.next != None:
            print(curNode.val, sep=" => ", end=" ")
            curNode = curNode.next

File: Linked Lists/LinkedList.java

public class LinkedList {

    Node headNode = new Node(0);
    Node tailNode = new Node(0);
    int size = 0;

    void append(int newVal) {

        // Create a new node
        Node newNode = new Node(newVal);

        // If linked list is empty, add newNode after head
        if (size == 0) {

            // Connect headNode and tailNode to newNode. Vice versa
            headNode.next = newNode;
            tailNode.prev = newNode;

            newNode.prev = headNode;
            newNode.next = tailNode;
        }
        else {

            // Get node before tailNode
            Node prevNode = tailNode.prev;

            // Connect prevNode to newNode. Vice versa
            prevNode.next = newNode;
            newNode.prev = prevNode;

            // Connect tailNode to newNode. Vice versa
            newNode.next = tailNode;
            tailNode.prev = newNode;
        }

        size++;
    }

    void delete(Node oldNode) {

        // OBJECTIVE: Delete oldNode from linked list

        // If oldNode is null, exit function
        if (oldNode == null) {return;}

        // Get neighboring nodes
        Node leftNode = oldNode.prev;
        Node rightNode = oldNode.next;

        // Connect neighboring nodes to each other
        leftNode.next = rightNode;
        rightNode.prev = leftNode;

        // Delete oldNode
        oldNode = null;
    }

    Node getHead() {

        // OBJECTIVE: Return node after headNode
        
        // If there's no node after headNode, exit function
        if (headNode.next == null) {return null;}

        return headNode.next;
    }

    int getSize() {

        // OBJECTIVE: Return linked list size
        return size;
    }

    void printList() {

        // OBJECTIVE: Print linked list

        // If linked list is empty, exit function
        if (headNode.next == null) {
            System.out.println("Linked list is empty!");
            return;
        }

        // Get headNode
        Node curNode = headNode.next;

        // Traverse linked list
        while (curNode.next != null) {
            System.out.print(curNode.val + " ");
            curNode = curNode.next;
        }
    }
}